File: common/src/main/java/com/raredev/vcspace/utils/PreferencesUtils.java
package com.raredev.vcspace.utils;

import android.content.SharedPreferences;
import com.raredev.vcspace.BaseApp;
import com.raredev.vcspace.res.R;

public class PreferencesUtils {

  public static SharedPreferences getDefaultPrefs() {
    return BaseApp.getInstance().getDefaultPref();
  }

  /*
   * Returns if the user wants to use dynamic theme(Material3)
   */
  public static boolean useDynamicColors() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_DYNAMIC_COLORS, true);
  }

  /*
   * Returns the user-selected font value
   */
  public static int getEditorTextSize() {
    return getDefaultPrefs().getInt(SharedPreferencesKeys.KEY_EDITOR_TEXT_SIZE, 14);
  }

  public static int getEditorLineHeight() {
    return getDefaultPrefs().getInt(SharedPreferencesKeys.KEY_EDITOR_LINEHEIGHT, 3);
  }

  public static int getEditorTABSize() {
    return Integer.valueOf(
        getDefaultPrefs().getString(SharedPreferencesKeys.KEY_EDITOR_TAB_SIZE, "4"));
  }

  /*
   * Returns the font the user wants to use in the editor
   */
  public static int getSelectedFont() {
    String selectedFont =
        getDefaultPrefs().getString(SharedPreferencesKeys.KEY_EDITOR_FONT, "firacode");
    return getFont(selectedFont);
  }

  public static boolean useStickyScroll() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_STICKYSCROLL, false);
  }

  public static boolean useFontLigatures() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_FONTLIGATURES, true);
  }

  public static boolean useWordWrap() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_WORDWRAP, false);
  }

  public static boolean lineNumbers() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_LINENUMBERS, true);
  }

  /*
   * Returns whether the user wants to use spaces instead of tabs(\t)
   */
  public static boolean useSpaces() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_USE_SPACES, true);
  }

  /*
   * Returns whether the user wants to quickly delete empty lines
   */
  public static boolean useDeleteEmptyLineFast() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_DELETE_EMPTY_LINE_FAST, true);
  }

  public static boolean useDeleteTabs() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_DELETETABS, true);
  }

  public static String getIndentationString() {
    String spaces = " ".repeat(getEditorTABSize());
    return useSpaces() ? spaces : "\t";
  }

  public static boolean showFilePath() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_FILE_PATH, true);
  }

  /*
   * Returns whether the user wants to automatically save the file
   */
  public static boolean autoSave() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_AUTO_SAVE, false);
  }

  /*
   * Returns the font selected by the user
   */
  private static int getFont(String selectedFont) {
    switch (selectedFont) {
      case "firacode":
        return R.font.firacode_regular;
      case "jetbrains":
        return R.font.jetbrains_mono;
      default:
        return R.font.firacode_regular;
    }
  }

  public static String getEncodingForOpening() {
    return getDefaultPrefs().getString(SharedPreferencesKeys.KEY_ENCODING_FOR_OPENING, "UTF-8");
  }

  public static boolean showHiddenFiles() {
    return getDefaultPrefs().getBoolean(SharedPreferencesKeys.KEY_SHOW_HIDDEN_FILES, true);
  }
}

File: common/src/main/java/com/raredev/vcspace/utils/PreferencesUtils.kt
/*
 * This file is part of Visual Code Space.
 *
 * Visual Code Space is free software: you can redistribute it and/or modify it under the terms of
 * the GNU General Public License as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Visual Code Space is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Visual Code Space.
 * If not, see <https://www.gnu.org/licenses/>.
 */

package com.raredev.vcspace.utils

import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatDelegate.*
import com.raredev.vcspace.app.BaseApplication
import com.raredev.vcspace.res.R

object PreferencesUtils {

  val prefs: SharedPreferences
    get() = BaseApplication.getInstance().getPrefs()

  // General
  val appTheme: Int
    get() = when (prefs.getString(SharedPreferencesKeys.KEY_THEME, "")) {
      "light" -> MODE_NIGHT_NO
      "dark" -> MODE_NIGHT_YES
      else -> MODE_NIGHT_FOLLOW_SYSTEM
    }

  val dynamicColors: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_DYNAMIC_COLORS, true)

  // Editor
  val textSize: Int
    get() = prefs.getInt(SharedPreferencesKeys.KEY_EDITOR_TEXT_SIZE, 14)

  val tabSize: Int
    get() = prefs.getString(SharedPreferencesKeys.KEY_EDITOR_TAB_SIZE, "4")!!.toInt()

  val selectedFont: Int
    get() = when (prefs.getString(SharedPreferencesKeys.KEY_EDITOR_FONT, "firacode")) {
      "firacode" -> R.font.firacode_regular
      "jetbrains" -> R.font.jetbrains_mono
      else -> R.font.firacode_regular
    }

  val stickyScroll: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_STICKY_SCROLL, false)

  val fontLigatures: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_FONT_LIGATURES, true)

  val wordWrap: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_WORDWRAP, false)

  val lineNumbers: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_LINE_NUMBERS, true)

  val useTab: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_USE_TAB, false)

  val deleteEmptyLineFast: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_DELETE_EMPTY_LINE_FAST, true)

  val deleteMultiSpaces: Boolean = prefs.getBoolean(SharedPreferencesKeys.KEY_DELETE_TABS, true)

  val identationString: String
    get() = if (useTab) "\t" else " ".repeat(tabSize)

  val autoSave: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_AUTO_SAVE, false)

  // File
  val showHiddenFiles: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_SHOW_HIDDEN_FILES, true)

  val isPythonFileExtracted: Boolean
    get() = prefs.getBoolean(SharedPreferencesKeys.KEY_PYTHON_FILE_EXTRACTED, false)
}
